rAF callback arg vs performance.now(), fixes my previous mistake

sidewayss opened the pull request:
The rAF callback timestamp is *not* identical to the value returned by `performance.now()`. I was wrong a few months back.  I'm don't remember exactly what I based that on. My only excuse is that I was recently back to programming after a few year hiatus.  Sorry.

Here I have corrected this mistake and added some detail. For an example of this in action: https://codepen.io/sidewayss/pen/XWorEoz

Open the console pane and see the results.  On Chrome I have seen the difference be >17ms or <0.3ms.  On Firefox it's always <3ms.  I can't get it to run on my older iOS device.  `performance.now()` always returns a value that is higher than (ahead of in time) the rAF callback timestamp, which makes sense once you understand how it works.  This stackoverflow puts this a proper context on this and has detail that should be integrated into this page too: https://stackoverflow.com/a/38360320/4941356

I'll take a crack at rewriting my changes to include some of that context and detail.

The rAF callback timestamp is coarsened to 2 decimals on Firefox and 3 on Chrome.  The page for `performance.now()` doesn't mention the way that Firefox has coarsened it to 2ms, which is a lot higher than the 100microseconds everyone else seems to be using, maybe that detail should be added there?  If anyone has more specifics on the coarsening of the rAF callback timestamp, they should probably be added here.

This answer on stackoverflow goes into some detail about the timestamps too: https://stackoverflow.com/a/57549862/4941356

sidewayss committed:
fixed the mistake I made in my last PR

The callback arg is not identical to the value returned by performance.now. I was wrong a few months back. I have corrected this mistake and added some detail.  More detail in the PR.

sidewayss committed:
fixed typo

sidewayss commented:
For me, the relevant functional questions are: What do I use as the "time value zero" in my animation?  Do I have to burn a frame just getting the initial time value on which to base the rest of my animation?  Or can I use the value returned by `performance.now()` prior to calling rAF and let the animation start working immediately?  

My answer is that both approaches can make sense, even with `performance.now()` being up to an entire frame ahead.
1) If your animation starts from the current property value, then go ahead and use `performance.now()` from prior to calling rAF.  The first frame will probably change the property value, though I cannot guarantee it without more information about the nature of these two values.  
2) If your animation jumps to a new value at "time zero", then you will probably want to burn that first frame setting that initial value, in which case you're not burning a frame at all.  That's the general idea, avoid burning a frame and thus delaying the start of the animation.

I don't know if you want to include any of that in this docs page, maybe as examples.  I'll leave it up to you.

sideshowbarker commented:
@yoavweiss, @igrigorik: if one of you (or both of you) can make any time to review this, that’d be great

sidewayss commented:
@sideshowbarker If this is going to take a while maybe I should submit a separate PR to simply back out the one word change I made in error, i.e. change "identical" back to "similar".  That would at least fix the error and then this PR could be about adding detail.  Let me know if you'd like me to do that.

sideshowbarker commented:
> @sideshowbarker If this is going to take a while maybe I should submit a separate PR to simply back out the one word change I made in error, i.e. change "identical" back to "similar". That would at least fix the error and then this PR could be about adding detail. Let me know if you'd like me to do that.

👍 Yes, please go ahead and do that

sidewayss commented:
#28659 has been created and it passes all the checks.

yoavweiss commented:
Thanks for filing this!! I think it'd be better to explain what the timestamp stands for, which I believe is the [end time of the rendering task](https://html.spec.whatwg.org/multipage/webappapis.html#event-loop-processing-model).

^^ @noamr

noamr commented:
This missing a key point in this timestamp, which is that it's shared between all windows that run on the same agent (i.e. all same origin windows, and more importantly same-origin iframes). This timestamp allows synchronizing an animation across frames that each one has its own rAF, which you can't do with `performance.now()`

sidewayss commented:
> I think it'd be better to explain what the timestamp stands for, which I believe is the [end time of the rendering task](https://html.spec.whatwg.org/multipage/webappapis.html#event-loop-processing-model).

@yoavweiss - Yes, I believe the stackoverflow links in the original post here refer to that concept but not to the spec directly. This is the reason that the timestamp always being behind `performance.now()` and the rendering time of properties set in the rAF callback is in the future.

`performance.now()` is a separate thing, but because it is based on the same clock, the same starting point in time, it is worth comparing the two.  `performance.now()` allows you to be a bit less behind the actual rendering time when you're inside the callback, and it allows you to avoid spending the first frame of the animation setting the property to the value it already contains.  It would be good to clarify the relationship between these two numbers somewhere.

> This missing a key point in this timestamp, which is that it's shared between all windows that run on the same agent (i.e. all same origin windows, and more importantly same-origin iframes). This timestamp allows synchronizing an animation across frames that each one has its own rAF, which you can't do with performance.now()

@noamr - Excellent point and should definitely be documented.  Why would you have separate animation callbacks running at the same time? Can you describe an example where it is necessary or desirable?

I'll take a stab at editing this into shape in the next couple of days.  Unless someone else would like to do it...

noamr commented:
> > This missing a key point in this timestamp, which is that it's shared between all windows that run on the same agent (i.e. all same origin windows, and more importantly same-origin iframes). This timestamp allows synchronizing an animation across frames that each one has its own rAF, which you can't do with performance.now()
> 
> @noamr - Excellent point and should definitely be documented. Why would you have separate animation callbacks running at the same time? Can you describe an example where it is necessary or desirable?
> 

Let's say you have an animation loop at 60FPS, and you have several 3p libraries that call `requestAnimationFrame` (e.g. an ad library, some animation in your chat widget, plus your main page). What differentiates this timestamp from calling `performance.now()` in a particular `rAF` callback is that it aligns with the general frame rate of the animation, and is not affected by those other callbacks or work tasks. So if you want some canvas sprite to move at an animation-aligned rate, this would give you a much more accurate representation than calling `performance.now()` in your particular `rAF`.

This also applies to when you have an animation in the outer window and an animation in an iframe and you want them to be in sync.

sidewayss commented:
Thanks for the examples, I'll probably quote you directly.

Can either of you provide detail about the "coarsening" of the timestamp?  In Chrome I always see 3 decimals and in Firefox I see 2 or fewer.

sidewayss commented:
What is the timing of setting a property value immediately prior to the first call to `requestAnimationFrame()`?  For example:
```
element.style.color = blue;
requestAnimationFrame(raf);
```
From what I've seen in recent experiments, that color setting renders prior to the first callback's settings, as if it's rendering in the frame represented by the first callback's timestamp.  Is that correct?  Or would you phrase it differently?  Or are there no guarantees and it might not render until the same frame as the first callback and thus never be rendered?

noamr commented:
> What is the timing of setting a property value immediately prior to the first call to `requestAnimationFrame()`? For example:
> 
> ```
> element.style.color = blue;
> requestAnimationFrame(raf);
> ```
> 
> From what I've seen in recent experiments, that color setting renders prior to the first callback's settings, as if it's rendering in the frame represented by the first callback's timestamp. Is that correct? Or would you phrase it differently? Or are there no guarantees and it might not render until the same frame as the first callback and thus never be rendered?

So the way this works depends on whether you're currently in updating the rendering or not.

e.g. setting the style prior to rendering:
```js

 // Element's style is changed before rendering, it will only take effect when rendering
setTimeout(() => {
  element.style.color = "blue";
  requestAnimationFrame(() => {
    / * Element will become blue in this frame */
  });
}, 1000);

// element will never be rendered blue, because it's overwritten before rendering
setTimeout(() => {
  element.style.color = "blue";
  requestAnimationFrame(() => {
    element.style.color = "red";
  });
}, 1000);

// element will be blue during the first frame, and red in the next frame
requestAnimationFrame(() => {
  element.style.color = "blue";
  requestAnimationFrame(() => {
    element.style.color = "red";
  });
};
```

> 
> Rephrased: Is the first call to the callback always in the next frame, never the current one? Does it always waits until the current frame has rendered to put the callback in the queue of functions to be called? But you can set a property value now and it will render in the current frame (most of the time).

`requestAnimationFrame` is always called in the next frame. But if you're not currently rendering, there is no "current" frame.

sidewayss commented:
OK, I don't know what I was seeing in my previous experiments, as I said, I was recently back from a hiatus.  I tested again and you are correct, there is no "current frame" in javascript, the rendering always happens in the next frame after the command is issued.  It makes sense.

And regarding the "coarsening", the reduction in precision?  There is plenty of info out there online about the coarsening of `performance.now()`, but not about `requestAnimationFrame()` and it appears that the two are coarsening in different ways.  This docs page currently says:
> This timestamp is a decimal number, in milliseconds, but with a minimal precision of 1ms (1000 µs).

There is a link at the bottom of the page to this blog post from 2012, which is pre-spectre and the reduction in timing precision everywhere: https://developer.chrome.com/blog/requestanimationframe-api-now-with-sub-millisecond-precision/
That link should maybe be removed?

As I said, I see only 3 decimals in Chrome, whereas `performance.now()` has lots of decimals.  In Firefox `performance.now` is an int and coarsened to 2ms, which is extreme, but the rAF timestamp has up to 2 decimal places.  Is there any more specific that we can be on this docs page?

noamr commented:
> OK, I don't know what I was seeing in my previous experiments, as I said, I was recently back from a hiatus. I tested again and you are correct, there is no "current frame" in javascript, the rendering always happens in the next frame after the command is issued. It makes sense.
> 
> And regarding the "coarsening", the reduction in precision? There is plenty of info out there online about the coarsening of `performance.now()`, but not about `requestAnimationFrame()` and it appears that the two are coarsening in different ways. This docs page currently says:

Yea the spec is currently not coarsening these and it varies between implementations.
Actually I need to follow up about this issue [here](https://github.com/whatwg/html/issues/7931).
Also see [this chromium bug](https://bugs.chromium.org/p/chromium/issues/detail?id=1422926).
This particular timestamp is not too Spectre sensitive because you can't invoke it in high frequency. But nonetheless it's a hole in the spec.

sidewayss commented:
Thanks for the info.  I don't have permission to see that chromium bug.

sidewayss commented:
Should I remove that link to the 2012 blog post?  Or does it still have valid info that is worth the link?

noamr commented:
> Should I remove that link to the 2012 blog post? Or does it still have valid info that is worth the link?

I think it's outdated.

sidewayss commented:
Tangential to the timing topic, this link in the same ending section of the page is also very old, and unless I hear otherwise I will assume that it should be deleted: https://robert.ocallahan.org/2010/08/mozrequestanimationframe-frame-rate_17.html

The other links are old, but the topics and info are still relevant.

Kaiido commented on files/en-us/web/api/window/requestanimationframe/index.md:
```diff
@@ -50,9 +50,17 @@ requestAnimationFrame(callback)
+    This argument also has its precision limited as illustrated by the limited number of
+    decimal places it uses, again varying by browser. In terms of the values themselves,
+    the {{domxref('performance.now()')}} value is always higher. The amount higher varies
+    by browser and momentary circumstances, but in some cases it can be almost as much as
```
It can actually be an arbitrarily large amount of time, e.g. in cases where a previous callback gets executed in the same frame and takes arbitrarily long time to execute.

```js
requestAnimationFrame((t1) => {
	while(performance.now() - t1 < 2000) {}
})
requestAnimationFrame((t1) => {
	console.log("late by", performance.now() - t1); // 2000.xx
})
```

sidewayss replied:
Yes it can.  I've also seen it happen with GPU overload when animating complex filters, outside of any javascript delay.  I will amend this to reflect your comment.  I hope to have some time over the next few days to get to it.  Based on other comments and reading that I did after filing this PR I'll be taking more of a chainsaw to this than a surgical knife, so I won't be making a tidy edit here in review-land, but I'll post a comment along with my push.  Thanks.

sidewayss replied:
@Kaiido - I misunderstood this comment last night, and though it is true and the code does log  differences > 2 seconds, this seems like an unusual setup, calling `requestAnimationFrame()` twice like that outside of a callback loop.
In practice, when would someone write code like this without looping in the callback?  This code (in [this codepen](https://codepen.io/sidewayss/pen/BavNmMq)) is the same concept but in a callback loop, and the timestamp is never much more than a frame behind `performance.now()`:
```
function raf(t) {
  const now = performance.now();
  console.log(t, now , t - now);
  while(performance.now() - t < 1000) {};
  if (x--)
    requestAnimationFrame(raf);  
}
let x = 3;
requestAnimationFrame(raf);
```

Kaiido replied:
Yes, having the same script require multiple times an animation frame is unusual and even a sign of bad code, however it's rather common to have multiple scripts / components that will require their own animation frame for various reasons, e.g. some scrolling effect + some continuous animation + some  event debouncing + etc. So it's actually pretty common to have `requestAnimationFrame` callbacks that end up being in concurrency with other callbacks.

sidewayss replied:
@Kaiido - Does the mousemove event throttling affect all operations that involve moving the mouse, such as the Window resize event? I have throttled the resize event in the past and I'm wondering if I didn't need to do so.

I ran your fiddle on Chrome and Firefox and mostly see what you describe.  What is "double frame buffering"?  Also, unfortunately, the results I see in Firefox are more erratic than what you describe.  Sometimes the direct call does not have the 2s delay time, and sometimes the timeout call has the 2s delay.

sidewayss replied:
Thanks for the explanations.  The Firefox discrepancies appeared several times as I fiddled, but they seem to appear rarely.  I'm on Windows, fwiw.

Kaiido commented:
But actually, for Window objects,  rAF timestamp nowadays seems to be always equal to [`document.timeline.currentTime`](https://developer.mozilla.org/en-US/docs/Web/API/AnimationTimeline/currentTime)?  
I guess there should at least be a mention to it in that article.

sidewayss commented:
Yes, [this codepen](https://codepen.io/sidewayss/pen/vYvOpya) illustrates that the rAF timestamp and `document.timeline.currentTime` are the same, except for occasional tiny variations in the first call to the callback.  I will mention that here.

sidewayss commented:
I just realized that `document.timeline.currentTime` offers an alternative to `performance.now()` if you want to avoid burning that first frame, and avoid setting your time-zero value inside the callback itself:
```
const zero = document.timeline.currentTime;
requestAnimationFrame(raf);
function raf(t) {
  animateValue(Math.abs(t - zero));
  requestAnimationFrame(raf);
}
```
It's a small difference, but it makes the delta between zero and the first frame's timeStamp close to a whole frame, and it uses a consistent set of time values instead of mixing different time sources. fwiw.

Testing it has been interesting because [codepen.io](https://codepen.io/sidewayss/pen/ExGVPRv) and [jsfiddle.net](https://jsfiddle.net/sidewayss/rbt70q8z/) can have what appears to be a multi-frame delay between the setting of `const zero` and the first call to the callback (codepen is more consistent and has a longer delay).  The [MDN playground](https://developer.mozilla.org/en-US/play?id=z3aFSQNqU4eRTHwKif8xoDgo85ovA%2FSxzNQ755hgDj%2BFXU68FOlZd8JfiZlZ9308Xz4dnQVk%2F56AHcep) does not have any such delays.  I assume that the delays are due to some kind of extra processing that those sites do, but there is never any such delay if you set `const zero = performance.now()`, so it's not that straightforward (see [this codepen](https://codepen.io/sidewayss/pen/rNoOxZm)).

Are there any caveats to this approach?  Does anyone know the reason for the multi-frame delay in codepen and jsfiddle?  Or should it simply be ignored?

Kaiido commented:
It depends on what you use it for, but if you base it as the initial time of an animation before entering rAF, you may actually miss the first frame: e.g. if you're starting your animation from an already animated document, you may have your callback fire as late as 16ms after and since `animateValue(0)` would never have been called your animation's initial position would never have been rendered. But once again that depends on what one actually wants.

Then, another caveat IMO, is that it's not available in web Workers.

sidewayss commented:
It's definitely worth mentioning, regardless of an example like the one I penned.  My assumption is that animations generally start from an element's current property value instead of jumping to a new value at the start of the animation.  But there are certainly cases where that's not true and you must set your time-zero inside the callback, as in the examples in this docs page.

sidewayss committed:
no new examples

sidewayss committed:
Merge branch 'main' into rAF

sidewayss commented:
I have pushed initial changes w/o any new examples just to get the descriptive text correct.  Sorry for the delay, but I have been busy and this requires enough thought that I didn't want to squeeze it in before bedtime.  Please review this basic text and then I can add a couple of different examples and you can edit or remove the ones you don't like.

sidewayss commented:
Note that in the two paragraphs (including the Warning) before the Syntax heading, this timestamp argument is also described/mentioned in some detail.  I don't know if all this info should be in one place or spread out, as it is now.  Let me know what you think.

sidewayss committed:
end time of prev frame's rendering

sidewayss committed:
Merge branch 'main' into rAF

sidewayss committed:
some sketchy examples

sidewayss committed:
Merge branch 'rAF' of https://github.com/sidewayss/content into rAF

sidewayss commented:
I have added some sketchy examples below the current example.  The `perfomance.now()` example is obviously the most expendable.  Let me know if/how you'd like to flesh out these examples, or if you prefer to remove them completely.  It would be ideal if the variable names in all the examples coincided, but I wanted to sketch this out and get some feedback before prettying it up.

sidewayss commented:
Something else that should probably be mentioned on this page:
If your callback is in a class, you must bind `this` when you call `requestAnimationFrame()`, or inside the callback `this` will be `undefined`.  These two fiddles illustrate:

[`requestAnimationFrame(this.animate);` `this` is undefined](https://jsfiddle.net/sidewayss/u7obcks6/1/)
[`requestAnimationFrame((t) => this.animate(t));` this works](https://jsfiddle.net/sidewayss/u7obcks6/2/)

If you set the event handler directly to the instance method, as in:
`btn.addEventListener("click", c.play, false);`
then in the callback `this` is the `<button>`, but I don't think that's worth mentioning.

sidewayss commented:
Here is a 3rd version of that fiddle that illustratesa difference between Chrome and Firefox, and a peculiarity in Chrome:
https://jsfiddle.net/sidewayss/u7obcks6/3/
_btw - all these fiddles require you to press the Play button to see the text fade in and see the console messages._

In Chrome only a few hundredths of a millisecond are passing between the first call to `requestAnimationFrame()` and the first call to the callback.  In Firefox it's over 100ms and varies +/- a few frames.  The tiny difference in Chrome is odd because it means there's a mismatch between `document.timeline.currentTime` and the callback timestamp.  Specifically here:
```
  play() {
    this.zero = document.timeline.currentTime;
    requestAnimationFrame((t) => this.firstFrame(t));
  }
  firstFrame(t) {
    console.log(t, this.zero, t - this.zero, document.timeline.currentTime - this.zero);
    this.animate(t);
  }
```
Note the rightmost two numbers in the console output.  Running a test page on my localhost I consistently see that `document.timeline.currentTime - this.zero` is the same as `t - this`, which means that when `document.timeline.currentTime` is accessed in `this.play()` it is a different value than when it is accessed in the `firstFrame()` callback, different by a tiny amount.  When I run this in the fiddle or from a page I hosted on my personal site, that 4th value in the console output is usually zero, but sometimes it matches my localhost.  If you want to run it locally you can get the source from my site here: https://sidewayss.com/test/test2.html
Regardless, the timestamp and `document.timeline.currentTime` are slightly off in Chrome here, and it's only noticeable because the callback is called immediately.

sidewayss committed:
meatier examplies

sidewayss commented:
Hi!  Is anyone out there? :) I just pushed some basic improvements to the examples I added two weeks ago.  IMO opacity is the best property for animation examples because it is the prototypical linear timing function property: values range from 0 to 1, no units to append.  It's also visually obvious and animates smoothly.  These 3 examples are all linear fade-ins of an element.

Also seems to me that there should be a "live" example for `requestAnimationFrame()`.  I was surprised there wasn't one.  [Here is a fiddle of a simple opacity fade in](https://jsfiddle.net/sidewayss/u7obcks6/) (that also debugs the various time sources we've been discussing here).  It would be easy to adapt that to a live example.  I did two live examples for another docs page, so I know what to do to create one (or more) if you'd like that.

IMO it would be good to merge the existing example with these three (or two if you don't like the `performance.now()` version).  I think opacity is better than transform as a hypothetical code snippet, it's certainly more compact.

Sorry this has gone so slowly, but with some feedback and edits from you all I can wrap it up this week.

noamr commented on files/en-us/web/api/window/requestanimationframe/index.md:
```diff
@@ -49,10 +50,14 @@ requestAnimationFrame(callback)
+    all windows that run on the same agent (i.e. all same origin windows, and more
+    importantly same-origin iframes), which allows synchronizing an animation across
+    multiple `requestAnimationFrame` callbacks. The timestamp value is also similar to
+    {{domxref('performance.now()')}}, but never the same value.
```
perhaps similar to calling performance.now() at the beginning of the callback?

sidewayss replied:
Based on our discussions here I was shying away from performance.now() here, but I can certainly add something to that effect.  I'll take your comment as a vote to keep the performance.now example too.

noamr replied:
I meant, instead of saying "similar to `performance.now()`" say "similar to calling `performance.now()` at the beginning of the callback" or so.

noamr replied:
Great. It was a nit, I'll try to find time for the other stuff. Thanks for working on this!

sidewayss committed:
performance.now() adjustment

sidewayss committed:
consolidate opening, callback info

sidewayss commented:
I had some time so I took a crack at consolidating the callback info that was spread across the opening and the description of the `callback` parameter.  I made it all more concise.  Hopefully you all will approve (I generally hesitate to do any major rewrites here at MDN docs, but this page needs it).

I left the red Warning section alone but, as I think I mentioned in a previous comment, I think it should be rewritten. I recently wrote some fresh raf() code to animate a progress bar during a time-consuming set of nested for loops.  [Here](https://stackoverflow.com/a/77123296/4941356) is my answer to my own question on stackoverflow, and [here it is at jsfiddle](https://jsfiddle.net/sidewayss/w5pt4vyL/4/). It reminded me that there are at least two basic approaches to raf():

- The one described in the Warning and in the examples, where the callback ends the animation by *not* calling `requestAnimationFrame()` when the time is up or the end value is reached, 
- The one in the fiddle, where by design you call `cancelAnimationFrame()` from outside the callback to end the animation. The callback has no knowledge of start and end times or values.  The callback runs in an indefinite loop controlled from the outside.

I would like to document both these approaches, not just the first one.  The way it's written, especially with that red Warning, makes it sound like there is only one approach to running raf(), only one kind of animation that exists.  What do you all think about that?

noamr commented:
Seems OK to me, but perhaps other people can look at the code examples to see if they're easy to understand.

sidewayss commented:
> ...perhaps other people can look at the code examples to see if they're easy to understand.

I have held off changes to the first, primary example here out of general caution for making major changes to these docs, but I still think it should be replaced by a fully functioning version of one of the 3 examples I added, like this fiddle: https://jsfiddle.net/sidewayss/u7obcks6/7/.  The current first example is long-winded, and aligning all the examples around a linear timing function that modifies `opacity` would be cleaner.

So there would be 3 examples of a linear timing function animation, one or more of them fully functional via a "play" link to the MDN Playground or a "live" example on github.  Then, if you deem it appropriate, there could be a fourth example that uses `cancelAnimationFrame()`, as described in my [previous comment](https://github.com/mdn/content/pull/28590/#issuecomment-1732387869).

If you all can decide on this stuff I can wrap it all up a.s.a.p.  Thanks!

sidewayss committed:
reorder examples, clarify choices

sidewayss commented:
I see that [this spec issue](https://github.com/whatwg/html/issues/7931) (mentioned above by @noamr) has been fixed [here](https://github.com/whatwg/html/pull/9823).  Any impact on these docs?

@sideshowbarker - any chance of reviewing this soon? If you can answer my previous questions, give me a little guidance, I can wrap this up quickly.  IMO all that's left to do is decide on the examples (I would like to replace the current example with 3 working examples based on the 3 that I added, and possibly add a 4th example that uses `cancelAnimationFrame()`).  But you should review all the changes I made, as I rewrote some paragraphs completely.

I just pushed a reordering of the examples because my understanding of first call to the callback has evolved recently.  Firefox has an approximately 100ms delay between the call to `requestAnimationFrame()` and the first call to the callback, which is a relatively large amount of time (6 frames at 60hz).  Should that be considered a bug?  Chrome calls the callback immediately, sometimes with a timestamp that is microseconds earlier than `document.timeline.currentTime` at the time of the call to `requestAnimationFrame()`, which is also strange enough to make me think it's worth reporting, though it's not reproducible 100% of the time, more like 25%.  Both these behaviors are illustrated in fiddles I linked to in comments above.

...just trying to keep this PR alive.  Thanks.

sideshowbarker commented on files/en-us/web/api/window/requestanimationframe/index.md:
```diff
@@ -9,35 +9,23 @@ browser-compat: api.Window.requestAnimationFrame
+browser that you wish to perform an animation. I requests that the browser call a
+user-supplied callback function prior to the next repaint.
+
+The frequency of calls to the callback function will generally match the display
+refresh rate, per the W3C recommendation. The most common refresh rate is 60hz,
+(60 cycles/frames per second), though 75hz, 120hz and 144hz are also widely used.
```
```suggestion
The frequency of calls to the callback function will generally match the display
refresh rate. The most common refresh rate is 60hz,
(60 cycles/frames per second), though 75hz, 120hz and 144hz are also widely used.
```

sideshowbarker committed:
Update files/en-us/web/api/window/requestanimationframe/index.md

sideshowbarker commented on files/en-us/web/api/window/requestanimationframe/index.md:
```diff
@@ -9,35 +9,23 @@ browser-compat: api.Window.requestAnimationFrame
+browser that you wish to perform an animation. I requests that the browser call a
+user-supplied callback function prior to the next repaint.
+
+The frequency of calls to the callback function will generally match the display
+refresh rate. The most common refresh rate is 60hz,
+(60 cycles/frames per second), though 75hz, 120hz and 144hz are also widely used.
```
```suggestion
The frequency of calls to the callback function will generally match the display
refresh rate. The most common refresh rate is 60hz,
(60 cycles/frames per second), though 75hz, 120hz, and 144hz are also widely used.
```

sidewayss replied:
I see you like the Oxford/Harvard/serial comma - I do too, but I figured that MDN used the standard American comma.  Is that defined somewhere in a style guide here?

sideshowbarker replied:
I’m not sure whether or not it is

sideshowbarker committed:
Update files/en-us/web/api/window/requestanimationframe/index.md

sideshowbarker commented on files/en-us/web/api/window/requestanimationframe/index.md:
```diff
@@ -9,35 +9,23 @@ browser-compat: api.Window.requestAnimationFrame
-> current time) to calculate how much the animation will progress in a frame,
-> **otherwise the animation will run faster on high refresh rate screens**.
-> Check the example below for a way to do this.
+background tabs or hidden {{ HTMLElement("iframe") }}s in order to improve
+performance and battery life.
+
+> **Note:** Your callback function must call `requestAnimationFrame()` again if
+> you want to animate another frame. `requestAnimationFrame()` is one-shot.
+
+> **Warning:** Be sure to always use the first argument (or some other method for
+> getting the current time) to calculate how much the animation will progress in
+> a frame, **otherwise the animation will run faster on high refresh rate screens**.
+> See the examples belows for ways to do this.
```
```suggestion
background tabs or hidden {{ HTMLElement("iframe") }}s, in order to improve
performance and battery life.

> **Note:** Your callback function must call `requestAnimationFrame()` again if
> you want to animate another frame. `requestAnimationFrame()` is one-shot.

> **Warning:** Be sure to always use the first argument (or some other method for
> getting the current time) to calculate how much the animation will progress in
> a frame — **otherwise the animation will run faster on high refresh-rate screens**.
> For ways to do that, see the examples below.
```

sideshowbarker committed:
Update files/en-us/web/api/window/requestanimationframe/index.md

sideshowbarker commented on files/en-us/web/api/window/requestanimationframe/index.md:
```diff
@@ -48,11 +36,22 @@ requestAnimationFrame(callback)
-    {{domxref("DOMHighResTimeStamp")}} similar to the one returned by
-    {{domxref('performance.now()')}}, indicating the point in time when
-    `requestAnimationFrame()` starts to execute callback functions.
+  - The function to call when it's time to update your animation for the next
+    repaint. This callback function is passed a single argument, a
+    {{domxref("DOMHighResTimeStamp")}} indicating end time of the previous frame's
+    rendering (based on the number of milliseconds since
+    [time origin](/en-US/docs/Web/API/DOMHighResTimeStamp#the_time_origin)).
+  - The timestamp is a decimal number, in milliseconds, but with a minimal
+    precision of 1ms. For `Window` objects (not `Workers`), it is equal to
+    {{domxref("document.timeline.currentTime")}}. This timestamp is shared
+    between all windows that run on the same agent (i.e. all same origin windows,
+    and more importantly same-origin iframes), which allows synchronizing
+    animations across multiple `requestAnimationFrame` callbacks. The timestamp
+    value is also similar to calling {{domxref('performance.now()')}} at the start
+    of the callback function, but it is never the same value.
+  - When multiple callbacks queued by `requestAnimationFrame()` begin to fire in
+    a single frame, each receives the same timestamp even though time has passed
+    during the computation of every previous callback's workload.
```
```suggestion
  - The function to call when it's time to update your animation for the next
    repaint. This callback function is passed a single argument: a
    {{domxref("DOMHighResTimeStamp")}} indicating the end time of the previous frame's
    rendering (based on the number of milliseconds since
    [time origin](/en-US/docs/Web/API/DOMHighResTimeStamp#the_time_origin)).
  - The timestamp is a decimal number, in milliseconds, but with a minimal
    precision of 1 millisend. For `Window` objects (not `Workers`), it is equal to
    {{domxref("document.timeline.currentTime")}}. This timestamp is shared
    between all windows that run on the same agent (that is, all same-origin windows,
    and more importantly same-origin iframes) — which allows synchronizing
    animations across multiple `requestAnimationFrame` callbacks. The timestamp
    value is also similar to calling {{domxref('performance.now()')}} at the start
    of the callback function, but it is never the same value.
  - When multiple callbacks queued by `requestAnimationFrame()` begin to fire in
    a single frame, each receives the same timestamp even though time has passed
    during the computation of every previous callback's workload.
```

sidewayss replied:
There's a typo here: "1 millisend" should be "1 millisecond".

sideshowbarker committed:
Update files/en-us/web/api/window/requestanimationframe/index.md

sideshowbarker committed:
Update files/en-us/web/api/window/requestanimationframe/index.md

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

sideshowbarker commented on files/en-us/web/api/window/requestanimationframe/index.md:
```diff
@@ -48,11 +36,22 @@ requestAnimationFrame(callback)
+    {{domxref("DOMHighResTimeStamp")}} indicating the end time of the previous frame's
+    rendering (based on the number of milliseconds since
+    [time origin](/en-US/docs/Web/API/DOMHighResTimeStamp#the_time_origin)).
+  - The timestamp is a decimal number, in milliseconds, but with a minimal
+    precision of 1 millisend. For `Window` objects (not `Workers`), it is equal to
+    {{domxref("document.timeline.currentTime")}}. This timestamp is shared
```
```suggestion
  - The timestamp is a decimal number, in milliseconds, but with a minimal
    precision of 1 millisecond. For `Window` objects (not `Workers`), it is equal to
    {{domxref("document.timeline.currentTime")}}. This timestamp is shared
```

sideshowbarker committed:
Update files/en-us/web/api/window/requestanimationframe/index.md

sideshowbarker commented:
Nice work. If any of the technical reviewers here end up having further feedback, it could be addressed in a follow-up PR
